### PR TITLE
Support multi-party queries in daml script

### LINKED ISSUE #7635: Support multi-party queries in DAML Script
In preparation for multi-party command submission on the ledger, we should also support multi-party queries in DAML Script.
This can probably be non-breaking my simply overloading `query` and friends using `IsParty` (technically adding more polymorphism is breaking but this is so hard to hit, that it doesn’t matter).

### PR DESCRIPTION
**Pull request by cocreature**
Small comment: This is technically not completely backwards compatible
since it messes with type inference. But to hit that you are doing
something sufficiently advanced that you should be able to add the
type annotation that is required

fixes #7635

changelog_begin

- [DAML Script] `query`, `queryContractId` and `queryContractKey` now
  accepts multiple parties using the `IsParties` abstraction used by
  `signatory`, `observer` and other fields. They will return all
  contracts for which any of the given parties is a stakeholder. Since
  `Party` is an instance of `IsParties`, this is fully backwards
  compatible.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by aherrmann-da — APPROVED** [review]
LGTM

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala [hunk 1]
```diff
@@ -370,6 +378,15 @@ object Converter {
       case _ => Left(s"Expected SParty but got $v")
     }
 
+  def toParties(v: SValue): Either[String, OneAnd[Set, Ref.Party]] =
+    v match {
+      case SList(FrontStackCons(x, xs)) =>
+        NonEmptyList(x, xs.toList: _*).traverse(toParty(_)).map(toNonEmptySet(_))
```
Comment (S11001001) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:384]: `OneAnd` is way more flexible for this kind of thing.

```suggestion
        OneAnd(x, xs).traverse(toParty(_)).map(oafp => /*oafp: OneAnd[FrontStack, Party]; some conversion here*/)
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala [hunk 2]
```diff
@@ -850,18 +873,19 @@ class JsonLedgerClient(
...
-      case _ => Future.unit
+  // Check that the parties in the token are a superset of the given parties.
+  private def validateTokenParties(parties: OneAnd[Set, Ref.Party], what: String): Future[Unit] = {
+    val tokenParties = Set(tokenPayload.readAs ++ tokenPayload.actAs: _*)
+    // First check is just for a nicer error message and would be covered by the second
+    if (tokenParties.isEmpty) {
+      Future.failed(
+        new RuntimeException(
+          s"Tried to $what as ${parties.toList.mkString(" ")} but token contains no parties."))
+    } else if (tokenParties == Set[String](parties.toList: _*)) {
```
Comment (S11001001) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:884]:
```suggestion
    } else if (tokenParties == parties.toSet.toSet[String]) {
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala (hunk 1 above)
Comment (cocreature) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:384]: ah good point, somehow it didn’t occur to me to use `OneAnd` for something other than `Set` here.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala (hunk 2 above)
Comment (cocreature) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:884]: Do you really mean `toSet` twice or `toList` followed by `toSet`?

Comment (cocreature) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:884]: Or maybe just `toSet` once?

Comment (S11001001) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:884]: The first `toSet` doesn't take a type parameter; the second one does, which is where you can do the widen that you're doing with natural transformation to `List` and then using covariance in the merged code. The second `toSet` will return the receiver with its type changed.

**Pull request merged**